### ISSUE TITLE
Uncertainty when reading RegConfig register

### ISSUE THREAD

**Issue by Dryvnt** [yupferris/rustendo64#35]
The `kseg0_cache_enabled` value is ambiguous when reading the RegConfig register.

There are many combinations of three bits that cause the bool to be false, but the bool type is not descriptive enough to tell which of these combinations has actually been stored in the RegConfig register, when the program wants to read it.

In addition, there are several other bits of RegConfig that aren't being handled whatsoever, which is obviously going to be a problem when reading.

I don't know if any code ever actually _reads_ from the RegConfig register, but this will be a problem if some code does.

I'm sorry for being zealous and making an issue for this, which might be an overreaction. I'm watching the series for the first time, and I noticed this problem just now. Just wanted to make sure this is properly addressed in orderly fashion.

**Comment by yupferris** [comment]
> There are many combinations of three bits that cause the bool to be false, but the bool type is not descriptive enough to tell which of these combinations has actually been stored in the RegConfig register, when the program wants to read it.

This is definitely true; I might make that a backing i32 or something. But are all of these combinations valid? Can you point to where I can learn more about _why_ these combinations exist and what they mean?

> In addition, there are several other bits of RegConfig that aren't being handled whatsoever, which is obviously going to be a problem when reading.

Pretty sure I deferred this specifically because we _weren't_ reading any of these values yet :)

**Comment by Dryvnt** [comment]
> Can you point to where I can learn more about why these combinations exist and what they mean?

I don't know _why_ the other combinations exist, but the datasheet explicitly shows what each combination _means_ in Table 5-6, page 145 (Which is referred to by the description of the Config Register, page 153). Why they decided it needed 3 bits to describe this is beyond my understanding.

> Pretty sure I deferred this specifically because we weren't reading any of these values yet :)

That makes sense. I must have missed it, my bad!

**Comment by yupferris** [comment]
> Table 5-6, page 145 (Which is referred to by the description of the Config Register, page 153)

Right, I'll keep this issue open as either a reminder to change this when we get to reading the reg, or if someone wants to fix the backing store and do a PR I'm cool with that as well.

**Comment by yupferris** [comment]
Reopening; this wasn't meant to be closed in the last PR.

**Comment by HarveyHunt** [comment]
> I don't know why the other combinations exist, but the datasheet explicitly shows what each combination means in Table 5-6, page 145 (Which is referred to by the description of the Config Register, page 153). Why they decided it needed 3 bits to describe this is beyond my understanding.

There are 3 bits so that implementation specific caching modes can be used (such as Write Through mode). An implementation **must** have support for at least CCA (Cache Coherency Attribute) 2 and 3, although I don't think that the N64's CPU implements any more.

**Comment by yupferris** [comment]
cool, thanks for the info :) I'm ignoring cache altogether for now; I have a feeling that's going to be ok for quite some time. In the future though as we hone in on the accuracy of the emu this is going to be awesome information to have, so thanks again :)

**Comment by Gravewish** [comment]
Yes, CCA ("cache coherency attribute") can be thought of as a selection of the "cache algorithm" to be used during memory accesses to that page. In particular, a value of 2 means "uncached" and 3 means "cacheable uncoherent". These are the only ones required to be implemented, and any other value depends on the processor implementation, so you'd have to check the programmer's guide to know the real meaning.

"See MIPS Run" chapter 6.2 p. 137 talks about this topic in passing.

The following chart comes from the official MIPS32 Architecture reference book, "MIPS32™ Architecture For Programmers Volume III: The MIPS32™ Privileged Resource Architecture":

![cca](https://cloud.githubusercontent.com/assets/25113102/23145653/d4abf316-f7c7-11e6-83db-534b8be5c297.png)

From what I can gather, N64's CPU only implements the 2 required CCAs (and everything else is equivalent to "Cacheable")

However, to make it clear why there are 3 bits, you can take a look at [interAptiv's documentation](https://imagination-technologies-cloudfront-assets.s3.amazonaws.com/documentation/MD00904-2B-interAptiv-SUM-02.01.pdf) as an example of a MIPS32 processor from 2016:

![cca_interaptiv](https://cloud.githubusercontent.com/assets/25113102/23145741/91836ba4-f7c8-11e6-8bcb-77f58d3b59ce.png)
